Nb feat/implement level db

nicobarra opened the pull request:
In the following PR I implemented an api that makes a CRUD on leveldb over a Person struct: 

* It has a router that creates the routes and executes the handlers. Also it has a decorator that takes the handler's returns and makes the http response
* Each handler has its own file with the implementation
* PersonStorage is separated from the handlers. It means that I used dependency injection with interfaces and instancing objects in order to make the code easier to change.
* It has inputs validations off request using go validator lib
* All is executed in `main.go` file

cagodoy commented on Makefile:
```diff
@@ -0,0 +1,3 @@
+dev:
+	@echo "[dev] Running":
+	@nodemon --exec go run cmd/api/main.go --signal SIGTERM 
```
maybe you can find a other solution for hot reloading implemented in Golang, because `nodemon` script use `nodejs`.

By the other side, if I clone the project now and then I will run `make dev` command the script will be failed because I have not installed `nodemon` on my computer (and it's not specified in a `package.json` file or in install 
 make command.

nicobarra replied:
Ok

cagodoy commented on pkg/api/create_person_handler.go:
```diff
@@ -0,0 +1,49 @@
+	"github.com/go-playground/validator/v10"
+)
+
+type CreatePersonHandlerInputs struct {
```
Maybe you must use the original `Person` type defined in the `Person` package because you can apply multiple metadata and you can centralize the person definition.

See the below photo:
 
<img width="683" alt="Screen Shot 2022-10-17 at 15 31 43" src="https://user-images.githubusercontent.com/8485620/196129822-eb3f773e-be00-4569-801d-fc4d699d54a7.png">

nicobarra replied:
Yeah, is better in that way. Done 👍

cagodoy commented on pkg/api/create_person_handler.go:
```diff
@@ -0,0 +1,49 @@
+
+	// prepare struct for body
+	body := &struct {
+		Person *person.Person `json:"person"`
```
Following your implementation y think that you use the `CreatePersonHandlerInputs` type definition

nicobarra replied:
But they are different, `CreatePersonHandlerInputs` contains the methods of `PersonStorage` struct, and `Person` is an struct with the fields required by the person object.
Maybe you say is better if I would pass `Person.person` instead of creating an initializing an struct with it?

cagodoy replied:
I meant that you define the type but I don't see that you use it anywhere else, in fact I did the search in the entire PR and you don't use it anywhere

cagodoy commented on pkg/api/router.go:
```diff
@@ -1,21 +1,100 @@
 type response struct {
-	Data interface{} `json:"data,omitempty"`
-	Meta interface{} `json:"meta,omitempty"`
+	Data  interface{} `json:"data,omitempty"`
```
remove this spaces

nicobarra replied:
Which spaces?

```// Define response structure
type response struct {
	Data  interface{} `json:"data,omitempty"`
	Meta  interface{} `json:"meta,omitempty"`
	Error interface{} `json:"error,omitempty"`
}
```

Here it is how I have those lines of codes right now.

cagodoy commented on pkg/api/create_person_handler.go:
```diff
@@ -0,0 +1,49 @@
+	"github.com/go-playground/validator/v10"
+)
+
+type CreatePersonHandlerInputs struct {
```
I think that you do not use this type of implementation on anywhere of PR

nicobarra replied:
Got it, already changed

cagodoy commented on pkg/api/create_person_handler.go:
```diff
@@ -0,0 +1,49 @@
+func (cp CreatePersonHandler) Invoke(ctx *handlerCtx) *handlerRes {
+	// Validate json schema data
+	validate := validator.New()
+	validate.Struct(ctx.ps)
```
Are you sure that the code is using this validation? 
Did you try each edge case using postman? 

In this case, I think that you must implement a Test because you can try all different input cases.
The same for all handlers

cagodoy commented on pkg/api/update_person_handler.go:
```diff
@@ -0,0 +1,50 @@
+	// Define structure for parse body
+	body := &struct {
+		Person *person.Person `json:"person"`
+	}{} // TODO(ca): What does this second '{}' means?
```
The `TODO` should have the initials of the writer. In this case `TODO(nb): ...`

Answering your question is because it's an inline definition, you are defining the struct and at the same time, you are initializing it. In short words, the second `{}` is an initialization with empty values for person=`nil`.

nicobarra replied:
Ok, good to know 👍

cagodoy commented on pkg/storage/db.go:
```diff
@@ -0,0 +1,23 @@
+}
+
+func New(filepath string) (*DB, error) {
+	db, err := leveldb.OpenFile(fmt.Sprintf("./%s", filepath), nil)
```
missing validation for empty value or invalid value for `filepath` param

cagodoy commented on pkg/storage/person/storage.go:
```diff
@@ -0,0 +1,136 @@
+
+func (s *ps) Create(p *person.Person) (*person.Person, error) {
+	// generate id for database
+
```
remove this linebreak

cagodoy commented on pkg/storage/person/storage.go:
```diff
@@ -0,0 +1,136 @@
+		return nil, err
+	}
+
+	var pperson *person.Person
```
use `person` or `p` it's more clear.

cagodoy commented on pkg/storage/person/storage.go:
```diff
@@ -0,0 +1,136 @@
+		return nil, err
+	}
+
+	var pperson *person.Person
```
use `person` or `p` it's more clear.

cagodoy commented on pkg/storage/person/storage.go:
```diff
@@ -0,0 +1,136 @@
+}
+
+func (s *ps) DeletePersonById(id string) (*person.Person, error) {
+	data, err := s.DB.DB.Get([]byte(id), nil)
```
missing validation for `id` param

cagodoy commented on pkg/storage/person/storage.go:
```diff
@@ -0,0 +1,136 @@
+	return pperson, err
+}
+
+func (s *ps) UpdatePersonById(id string, p *person.Person) (*person.Person, error) {
```
mising validations for params

cagodoy commented on pkg/storage/person/storage.go:
```diff
@@ -0,0 +1,136 @@
+	return data, nil
+}
+
+func (s *ps) Create(p *person.Person) (*person.Person, error) {
```
missing validations for params

cagodoy commented on pkg/api/router.go:
```diff
@@ -1,21 +1,100 @@
+
+		// Prepare response
+		res := response{
+			Meta: map[string]interface{}{"status_code": statusCode},
```
its neccesary define its the `json` response should be use `cammelCase` or `snakeCase`

nicobarra commented:
Last changes:
* Fixed requested improvements from @cagodoy (they are marked with a 👍)
* Created internal validation func's in personstorage for validating params
* Added assertion in personstorage Create func for failing if the person to create already exists
